### PR TITLE
Fix kubectl describe for pods with controllerRef

### PR DESCRIPTION
**Pull request by deyuanix**
**What this PR does / why we need it**:

kubectl describe doesn't take controllerRef into consideration, resulting confusing result. e.g. if we have two replicaset with the same selector, one with 1 replica and the other 2 replicase, then both replicaset will show 3 running pods.

```sh
$ kubectl describe rs replicaset-2
Name:           replicaset-2      
Namespace:      default
Selector:       environment=prod
Labels:         environment=prod
Annotations:    <none>
Replicas:       2 current / 2 desired
Pods Status:    3 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:       environment=prod
  Containers:
   created-from-replicaset:
    Image:              nginx
    Port:               
    Environment:        <none>
    Mounts:             <none>
  Volumes:              <none>
Events:
  FirstSeen     LastSeen        Count   From                    SubObjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                    -------------   --------        ------                  -------
  5m            5m              1       replicaset-controller                   Normal          SuccessfulCreate        Created pod: replicaset-2-39szb
  5m            5m              1       replicaset-controller                   Normal          SuccessfulCreate        Created pod: replicaset-2-470jr
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

xref #24946

**Special notes for your reviewer**:

**Release note**:

```release-note
Fix kubectl describe for pods with controllerRef 
```

### CODE REVIEW

**Comment by spiffxp** [comment]
@k8s-bot ok to test

**Comment by janetkuo** [comment]
Thanks, we should fix it in `kubectl get` too. 

@ddysher: I'm curious why you created 2 RSes with the same selector. Did you do it by accident or is it a special use case for you?

**Comment by ddysher** [comment]
@janetkuo not really, we want to use this feature to find pod->controller, so I'm simply trying out controllerRef.

Where do I need to fix for `kubectl get`; get doesn't counts pods, maybe there's other options I need to consider?
```
$ kubectl get rs     
NAME           DESIRED   CURRENT   READY     AGE
replicaset-1   1         1         1         2s
replicaset-2   2         2         2         31s

$ kubectl get pods     
NAME                 READY     STATUS    RESTARTS   AGE
replicaset-1-z3d05   1/1       Running   0          11s
replicaset-2-h5tll   1/1       Running   0          40s
replicaset-2-mmnl3   1/1       Running   0          40s
```

**Comment by janetkuo** [comment]
@ddysher you're right, `get` only dump information from controller status, so we only need to fix `describe`.

File: pkg/printers/internalversion/describe.go [hunk 1]
```diff
@@ -2650,6 +2653,18 @@ func getPodStatusForController(c coreclient.PodInterface, selector labels.Select
 	return
 }
 
+// isControllerRef returns true if a pod is managed by a controller or if owner
+// reference is not set; otherwise, return false.
```
Comment (ddysher) [inline pkg/printers/internalversion/describe.go]: I'm returning `true` here mainly for backward compatibility; didn't realize orphaned pod case.  If a pod is orphaned, will we be able to select it?  All pods passed in there are selected via controller's selector.

Comment (janetkuo) [inline pkg/printers/internalversion/describe.go]: An orphan pod = a pod that doesn't have a controllerRef specified. ControllerRef is a lock when one child is selected by multiple parents via selector. That's to say, a controller can select a pod that it doesn't own, but the controller won't try to manage that pod. 

A pod can be orphaned when its controller is deleted non-cascadingly. Other existing controllers with matching selector will try to adopt it, and eventually at most one controller will win.

File: pkg/printers/internalversion/describe_test.go [hunk 2]
```diff
@@ -1398,3 +1398,74 @@ URL:	http://localhost
...
+					{Image: "mytest-image:latest"},
+				},
+			},
+			Status: api.PodStatus{
+				Phase: api.PodRunning,
+			},
+		})
+	d := ReplicationControllerDescriber{f}
+	out, err := d.Describe("foo", "bar", printers.DescriberSettings{ShowEvents: false})
+	fmt.Println(string(out))
```
Comment (enisoc) [inline pkg/printers/internalversion/describe_test.go]: It's better to send this to `t.Log()` so it only shows if the test fails. Also it should be already a `string` without casting, right?

File: pkg/printers/internalversion/describe.go [hunk 3]
```diff
@@ -2654,6 +2657,16 @@ func getPodStatusForController(c coreclient.PodInterface, selector labels.Select
 	return
 }
 
+// isControlledBy returns true if pod is controlled by given controller or if
+// owner reference is not set; otherwise, return false.
+func isControlledBy(pod api.Pod, name string, uid types.UID) bool {
+	controllerRef := controller.GetControllerOf(&pod)
+	if controllerRef != nil && (controllerRef.Name != name || controllerRef.UID != uid) {
```
Comment (janetkuo) [inline pkg/printers/internalversion/describe.go]: Yes comparing `uid` is enough.

File: pkg/printers/internalversion/describe_test.go (hunk 2 above)
Comment (ddysher) [inline pkg/printers/internalversion/describe_test.go]: Thanks for pointing this out.  I'll just remove it since we already print it in error case.

**Comment by ddysher** [comment]
@janetkuo @enisoc comment addressed.

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
```
I0515 05:33:49.549]   /go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go is out of date. Please run hack/update-staging-client-go.sh
I0515 05:33:49.549] FAILED   hack/make-rules/../../hack/verify-staging-client-go.sh	101s
```

File: pkg/printers/internalversion/describe.go [hunk 4]
```diff
@@ -2633,13 +2633,18 @@ func printReplicaSetsByLabels(matchingRSs []*versionedextension.ReplicaSet) stri
...
+func getPodStatusForController(c coreclient.PodInterface, selector labels.Selector, uid types.UID) (running, waiting, succeeded, failed int, err error) {
 	options := metav1.ListOptions{LabelSelector: selector.String()}
 	rcPods, err := c.List(options)
 	if err != nil {
 		return
 	}
 	for _, pod := range rcPods.Items {
+		controllerRef := controller.GetControllerOf(&pod)
+		// Skip pods that are orphans or owned by other controllers.
+		if controllerRef != nil && controllerRef.UID != uid {
```
Comment (enisoc) [inline pkg/printers/internalversion/describe.go]: This still doesn't skip orphans (`controllerRef = nil`).

It should be:

```go
if controllerRef == nil || controllerRef.UID != uid {
  continue
}
```

Please also add an orphan to the unit test to verify this works.

**Comment by ddysher** [comment]
@janetkuo @enisoc updated

File: pkg/printers/internalversion/describe.go (hunk 4 above)
Comment (ddysher) [inline pkg/printers/internalversion/describe.go]: thx!  I didn't look closely...

**Comment by ddysher** [comment]
@janetkuo seems e2e is failing for different reasons

**Comment by ddysher** [comment]
@lavalamp

**Comment by janetkuo** [comment]
Please squash commits

**Comment by ddysher** [comment]
@janetkuo done

**Comment by ddysher** [comment]
rebased

ping @lavalamp

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

**Comment by enisoc** [comment]
@ddysher Sorry about the delay in getting someone who can approve this. To unblock the actual fix for kubectl, you could move your `BoolPtr()` function to be a package-local `boolPtr()` in the package where you need it. That would limit the scope of this PR to files that @janetkuo can approve.

**Comment by ddysher** [comment]
@janetkuo @enisoc updated

**Pull request merged**